File: NeetCode/Binary_Search/Capacity_To_Ship_Packages_Within_D_Days/Capacity_To_Ship_Packages_Within_D_Days.cpp
// Capacity_To_Ship_Packages_Within_D_Days.cpp
//
// Compile locally with:
// clang++ Capacity_To_Ship_Packages_Within_D_Days.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int shipWithinDays(vector<int> &weights, int days)
    {
        int left = 0;
        int right = 0;
        for (auto &tmp : weights)
        {
            left = max(left, tmp);
            right += tmp;
        }

        int d(1), mid(0), curr(0);
        while (left < right)
        {
            mid = left + (right - left) / 2;
            curr = 0;
            d = 1;
            for (auto &weight : weights)
            {
                if (curr + weight > mid)
                {
                    curr = weight;
                    d++;
                }
                else
                {
                    curr += weight;
                }
            }
            if (d <= days)
            {
                right = mid;
            }
            else
            {
                left = mid + 1;
            }
        }

        return left;
    }
};

int main()
{
    cout << "Running Capacity_To_Ship_Packages_Within_D_Days.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: NeetCode/Binary_Search/Capacity_To_Ship_Packages_Within_D_Days/capacity__to__ship__packages__within__d__days.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        left = max(weights)
        right = sum(weights)

        while left < right:

            mid = left + (right - left) // 2
            # see if can be done with mid weights in x days
            d = 1
            currWeight = 0
            for weight in weights:
                if currWeight + weight > mid:
                    currWeight = weight
                    d += 1
                else:
                    currWeight += weight
            if d <= days:
                right = mid
            else:
                left = mid + 1
        return left


if __name__ == "__main__":
    sol = Solution()
    weights = [2, 4, 6, 1, 3, 10]
    days = 4
    print(sol.shipWithinDays(weights, days))
    print("Running Solution...")
